### DISCUSSION TITLE
Pass Parameter to Query String from onInit or onRender

### DISCUSSION THREAD

**Discussion by samant-rags** [gapitio/gapit-htmlgraphics-panel#106] [General] [answered]
if i want to set a parameter value in the Query tab from  onInit or onRender  method, is it possible ?

if not, can we use a hidden variable (dashboard settings-->variables--> hide )  and then use that in the query BUT set value of that variable from onInit or onRender  method ?

Thanks in Advance 

![image](https://user-images.githubusercontent.com/5258222/178145183-ed0f7a02-f678-4a87-b2c8-97fae415bc96.png)

**Comment by ZuperZee** [comment, marked as answer]
Updating a Grafana variable is probably the best.
[Example on how to update and read Grafana variables.](https://gapit-htmlgraphics-panel.gapit.io/docs/examples/update-grafana-variable-example/).

**Reply by samant-rags** [reply]
while this is acceptable , 
was wondering if we can manipulate/update the query somehow instead of proxying it via Variables :)

Thanks Zuper as always !

**Reply by ZuperZee** [reply]
You could also write the query directly in the onInit or onRender code.
An example on [how to export to csv](https://gapit-htmlgraphics-panel.gapit.io/docs/examples/export-to-csv/). Instead of converting it to CSV you could just read the value from the response.

I made a short code example on how you could approach using influx2.
![image](https://user-images.githubusercontent.com/39485579/178495222-07532b33-db6b-4b47-b188-5addac948b01.png)
`HTML`
```html
<div id="root"></div>
```

`onRender`
```js
const root = htmlNode.querySelector("#root");

function createList(times, values) {
  const maxElements = Math.min(times.length, 5); // Limit to 5 elements

  const newChildren = Array.from({ length: maxElements }).map((_, i) => {
    const value = values[i];
    const time = times[i];
    const date = new Date(time).toISOString();

    const div = document.createElement("div");
    div.textContent = `${date}: ${value}`;
    return div;
  });

  root?.replaceChildren(...newChildren);
}

function main() {
  const query = `from(bucket: "bucket")
  |> range(start: v.timeRangeStart, stop: v.timeRangeStop)
  |> filter(fn: (r) => r["_measurement"] == "go_memstats_alloc_bytes")
  |> aggregateWindow(every: v.windowPeriod, fn: mean, createEmpty: false)
  |> yield(name: "mean")`
  // const refId = "A";
  // const datasource = {
  //   "type": "influxdb",
  //   "uid": "J4AWqX6nk"
  // };
  const [{ refId, datasource }] = data.request.targets; // Comment out this and uncomment the above with the correct values
  const { from, to } = data.timeRange;

  const body = {
    queries: [
      {
        datasource,
        query,
        refId,
        intervalMs: 0,
        maxDataPoints: 1e9, // 1E9 = 1_000_000_000 (Max allowed 1E20)
      },
    ],
    from: String(from.valueOf()),
    to: String(to.valueOf()),
  };

  fetch("/api/ds/query", {
    headers: {
      "cache-control": "no-cache",
      "content-type": "application/json",
    },
    body: JSON.stringify(body),
    method: "POST",
  })
    .then((res) => res.json())
    .then((data) => {
      const [
        {
          frames: [
            {
              data: { values },
            },
          ],
        },
      ] = Object.values(data.results);

      createList(...values)
    })
    .catch(console.error);
}

main();
```

You can also remove the whole query if you change the variables `refId` and and `datasource` in the `onRender` code (_See the comment in the `onRender` code_).

**Reply by samant-rags** [reply]
super  Zuper !   Thanks a lot for the help !   Was busy with other things - will try this and see

**Reply by samant-rags** [reply]
Thanks Zuper , this works well . However, if the div content is dynamically set
 and overflows, then getting an external scroll bar based on overFlow settings.
Wondering if it can be like the natural scrollbar of Panel (shown in 3rd image) ?

![image](https://user-images.githubusercontent.com/5258222/181491045-d363f872-9042-4a1e-82e1-7980375d3a5a.png)



![image](https://user-images.githubusercontent.com/5258222/181490882-dc560fd9-5f95-431f-8b81-9e382160edba.png)

**Reply by samant-rags** [reply]
like this when the html content  is static

![image](https://user-images.githubusercontent.com/5258222/181491871-8012f314-7020-4cad-bf9b-00cd4f2f2177.png)

**Reply by ZuperZee** [reply]
Created an issue for this https://github.com/gapitio/gapit-htmlgraphics-panel/issues/109
Not sure when a new release will be

**Reply by ZuperZee** [reply]
@samant-rags, just to let you know, the scrollbar has been added in [v1.2.0](https://gapit-htmlgraphics-panel.gapit.io/docs/release-notes/v2.1.0/)

**Reply by samant-rags** [reply]
Thanks a lot for this Zuper  :)  

Wondering if anything can be done for this - https://github.com/gapitio/gapit-htmlgraphics-panel/issues/65  ?

**Reply by samant-rags** [reply]
hi Zuper ,
Not sure if this is  issue with the plugin or my custom code.. but on non-black/default background seeing this issue. 
If i make the scroll auto , its fine 

![image](https://user-images.githubusercontent.com/5258222/184320573-5bcb6bda-93e6-4255-ad90-cf079ad5c7c2.png)

**Reply by ZuperZee** [reply]
> hi Zuper , Not sure if this is issue with the plugin or my custom code.. but on non-black/default background seeing this issue. If i make the scroll auto , its fine
> 
> ![image](https://user-images.githubusercontent.com/5258222/184320573-5bcb6bda-93e6-4255-ad90-cf079ad5c7c2.png)

Do you have an example code?

**Reply by samant-rags** [reply]
The HTML and onRender script 
-------------------
 `<div class="p-4 max-w-md bg-white rounded-lg border shadow-md sm:p-8 dark:bg-gray-800 dark:border-gray-700">
    <div class="flex justify-between items-center mb-4">
        <h5 class="text-xl font-bold leading-none text-gray-900 dark:text-white">Recent Orders</h5>
        <a href="#" class="text-sm font-medium text-blue-600 hover:underline dark:text-blue-500">
            View all
        </a>
   </div>
   <div class="flow-root">
        <ul role="list" class="  dark:divide-gray-700">
			<div id="listRoot" class="divide-y divide-gray-200 overflow:scroll" ></div>   
        </ul>
   </div>
</div>
 
<link href="https://unpkg.com/tailwindcss@^2/dist/tailwind.min.css" rel="stylesheet">
`
-----------------

	displayList();
	
 	function displayList() {
	
		let shadowRootDiv =htmlNode.getElementById("listRoot");
		let productListLength = data.series[0].fields[0].values.length;
		let productListTag='';
		 
		for (i=0; i < productListLength ; i++){ 
			console.log(data.series[0].fields[0].values.get(i));
			productListTag += mergeDataWithListElement(data.series[0].fields[0].values.get(i),
													   data.series[0].fields[1].values.get(i),
													   Math.floor(Math.random() * 1001));
		} 

		if(shadowRootDiv)
		 htmlNode.getElementById('listRoot').innerHTML = productListTag;
		else
			console.log("listRoot divID not found..");		
			
	}
	
	
	function mergeDataWithListElement (productId,productDesc,productPrice){
	
		let dataRespTemplate = `
							 <li class="py-3 sm:py-4 ">
								<div class="flex items-center space-x-4"   >
									<div class="flex-shrink-0">
										<img class="w-15 h-12 rounded-full scale-50 hover:scale-75 ease-in duration-500" src="https://staticimg.titan.co.in/Tanishq/Catalog/${productId}.jpg" alt="click ">
									</div> 
									<div class="flex-1 min-w-0">
										<p class="text-sm font-medium text-gray-900 truncate dark:text-white">
											${productDesc}
										</p>
										<p class="text-sm text-gray-500 truncate dark:text-gray-400">
											${productId}
										</p>
									</div>
									<div class="inline-flex items-center text-base font-semibold text-gray-900 dark:text-white">
										$ ${productPrice}
									</div>
								</div>
							</li>
 						` ;
	
			return dataRespTemplate ; 
	}

**Reply by samant-rags** [reply]
and the datsource= postgresSQl  with query

select productid ,productdesc from customerproducts
where sourceid ='hawkeye'

**Reply by ZuperZee** [reply]
Doesn't seem like the "view all" link is working and the CSS is missing

**Reply by samant-rags** [reply]
> Doesn't seem like the "view all" link is working and the CSS is missing

Sorry View all is not imp here..  i was trying to highlight the  outerDiv not streching as the list grows due to scroll.

I added some CSS for the outdiv and its fine for now

**Reply by ZuperZee** [reply]
I can only see the onRender code
![image](https://user-images.githubusercontent.com/39485579/184470957-1c6d549e-e02a-46c2-baf2-d5ac1551b266.png)

Do you have the code for CSS and HTML? (Exporting the entire panel-options would be better)

**Reply by samant-rags** [reply]
hi Zuper
PFB the entire panel options  export  json 
--------------

{
    "calcsMutation": "standard",
    "reduceOptions": {
        "calcs": [
            "lastNotNull",
            "last",
            "firstNotNull",
            "first",
            "min",
            "max",
            "mean",
            "sum",
            "count",
            "range",
            "delta",
            "step",
            "diff",
            "logmin",
            "allIsZero",
            "allIsNull",
            "diffperc"
        ]
    },
    "add100Percentage": true,
    "centerAlignContent": true,
    "overflow": "visible",
    "useGrafanaScrollbar": true,
    "SVGBaseFix": true,
    "codeData": "{\n  \"text\": \"Random text\"\n}",
    "rootCSS": "",
    "css": " ",
    "html": "<div class=\"p-4 bg-white max-w-md full-panel rounded-lg border shadow-md sm:p-8 \">\n    <div class=\"flex justify-between items-center mb-4\">\n        <h5 class=\"text-xl font-bold leading-none text-gray-900 dark:text-white\">Recent Orders</h5>\n        <a href=\"#\" class=\"text-sm font-medium text-blue-600 hover:underline dark:text-blue-500\">\n            View all\n        </a>\n   </div>\n   <div class=\"flow-root bg-white\">\n        <ul>\n\t\t\t<div id=\"listRoot\" class=\"divide-y divide-gray-200 \" ></div>   \n        </ul>\n   </div>\n</div>\n \n<link href=\"https://unpkg.com/tailwindcss@^2/dist/tailwind.min.css\" rel=\"stylesheet\">\n\n",
    "renderOnMount": true,
    "onRender": "\n\tdisplayList();\n\t\n \tfunction displayList() {\n\t\n\t\tlet shadowRootDiv =htmlNode.getElementById(\"listRoot\");\n\t\tlet productListLength = data.series[0].fields[0].values.length;\n\t\tlet productListTag='';\n\t\t \n\t\tfor (i=0; i < productListLength ; i++){ \n\t\t\tconsole.log(data.series[0].fields[0].values.get(i));\n\t\t\tproductListTag += mergeDataWithListElement(data.series[0].fields[0].values.get(i),\n\t\t\t\t\t\t\t\t\t\t\t\t\t   data.series[0].fields[1].values.get(i),\n\t\t\t\t\t\t\t\t\t\t\t\t\t   Math.floor(Math.random() * 1001));\n\t\t} \n\n\t\tif(shadowRootDiv)\n\t\t htmlNode.getElementById('listRoot').innerHTML = productListTag;\n\t\telse\n\t\t\tconsole.log(\"listRoot divID not found..\");\t\t\n\t\t\t\n\t}\n\t\n\t\n\tfunction mergeDataWithListElement (productId,productDesc,productPrice){\n\t\n\t\tlet dataRespTemplate = `\n\t\t\t\t\t\t\t <li class=\"py-3 sm:py-4 \">\n\t\t\t\t\t\t\t\t<div class=\"flex items-center space-x-4 \"   >\n\t\t\t\t\t\t\t\t\t<div class=\"flex-shrink-0\">\n\t\t\t\t\t\t\t\t\t\t<img class=\"w-15 h-12 rounded-full scale-50 hover:scale-75 ease-in duration-500\" src=\"https://staticimg.titan.co.in/Tanishq/Catalog/${productId}.jpg\" alt=\"click \">\n\t\t\t\t\t\t\t\t\t</div> \n\t\t\t\t\t\t\t\t\t<div class=\"flex-1 min-w-0\">\n\t\t\t\t\t\t\t\t\t\t<p class=\"text-sm font-medium text-gray-900 truncate dark:text-white\">\n\t\t\t\t\t\t\t\t\t\t\t${productDesc}\n\t\t\t\t\t\t\t\t\t\t</p>\n\t\t\t\t\t\t\t\t\t\t<p class=\"text-sm text-gray-500 truncate dark:text-gray-400\">\n\t\t\t\t\t\t\t\t\t\t\t${productId}\n\t\t\t\t\t\t\t\t\t\t</p>\n\t\t\t\t\t\t\t\t\t</div>\n\t\t\t\t\t\t\t\t\t<div class=\"inline-flex items-center text-base font-semibold text-gray-900 dark:text-white\">\n\t\t\t\t\t\t\t\t\t\t$ ${productPrice}\n\t\t\t\t\t\t\t\t\t</div>\n\t\t\t\t\t\t\t\t</div>\n\t\t\t\t\t\t\t</li>\n \t\t\t\t\t\t` ;\n\t\n\t\t\treturn dataRespTemplate ; \n\t}",
    "dynamicHtmlGraphics": false,
    "dynamicData": false,
    "dynamicFieldDisplayValues": false,
    "dynamicProps": false,
    "panelupdateOnMount": true,
    "onInitOnResize": false,
    "onInit": " "
}

**Reply by ZuperZee** [reply]
Doesn't seem like it was copied correctly. If you surround the code with triple `
Example:
```
{
    "calcsMutation": "standard",
    "reduceOptions": {
        "calcs": [
            "lastNotNull",
            "last",
            "firstNotNull",
            "first",
            "min",
            "max",
            "mean",
            "sum",
            "count",
            "range",
            "delta",
            "step",
            "diff",
            "logmin",
            "allIsZero",
            "allIsNull",
            "diffperc"
        ]
    },
    "add100Percentage": true,
    "centerAlignContent": true,
    "overflow": "visible",
    "useGrafanaScrollbar": true,
    "SVGBaseFix": true,
    "codeData": "{\n  \"text\": \"Random text\"\n}",
    "rootCSS": "",
    "css": "* {\n  font-family: Open Sans;\n}\n\n.box {\n  border: solid #555 2px;\n  border-radius: 10px;\n  padding: 10px 20px;\n}\n",
    "html": "<div style=\"text-align: center;\">\n  <div class=\"box\" id=\"htmlgraphics-text\"></div>\n  <br />\n  <div class=\"box\" id=\"htmlgraphics-value\"></div>\n</div>\n",
    "renderOnMount": true,
    "onRender": "// Sets the value from the first series on every refresh\nconst htmlgraphicsValue = htmlNode.getElementById('htmlgraphics-value');\n\nif (htmlgraphicsValue) {\n  const valueField = data.series[0]?.fields[1];\n  if (valueField) {\n    const length = valueField.values.length;\n    htmlgraphicsValue.textContent = valueField.values.get(length - 1);\n  } else {\n    htmlgraphicsValue.textContent = \"No data\"\n  }\n}\n",
    "dynamicHtmlGraphics": false,
    "dynamicData": false,
    "dynamicFieldDisplayValues": false,
    "dynamicProps": false,
    "panelupdateOnMount": true,
    "onInitOnResize": false,
    "onInit": "// Sets the text from customProperties\nconst htmlgraphicsText = htmlNode.getElementById('htmlgraphics-text');\n\nif (htmlgraphicsText) {\n  htmlgraphicsText.textContent = customProperties.text;\n\n  // Change the text color based on the theme\n  if (theme.isDark) {\n    htmlgraphicsText.style.color = 'green';\n  } else {\n    htmlgraphicsText.style.color = 'red';\n  }\n}\n"
}
```

If you could. Could you try creating an issue with a minimal example of how to reproduce it? _As I'm currently not able to reproduce it locally_
